crypto/err/err.c

```diff
@@ -742,9 +740,6 @@ ERR_STATE *ERR_get_state(void)
             CRYPTO_THREAD_set_local(&err_thread_local, NULL);
             return NULL;
         }
-
-        /* Ignore failures from these */
-        OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL);
```

Comment:
Why isn't this required any more? (it seems to be the only place in the library where we set this option)

Reply:
It's just that ERR functions may be called down the line from `OPENSSL_init_crypto()`, so this causes an init lockup.

Reply:
We *might* want to consider adding a `OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL)` in the error printing functions, or the error string retrieval functions.  That's where they are actually needed, and those are called explicitly by the application

Reply:
But if you don't do this, I can't see how the crypto strings get loaded.

Reply:
Yeahok...  so I gotta ask, since we get them loaded unconditionally here, what's the reason we pretend that it's optional to do so?  In other words, why not this?

``` diff
diff --git a/crypto/init.c b/crypto/init.c
index d8acfa03ec..afbb5a6872 100644
--- a/crypto/init.c
+++ b/crypto/init.c
@@ -615,15 +615,14 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
     if (!RUN_ONCE(&load_crypto_nodelete, ossl_init_load_crypto_nodelete))
         return 0;
 
-    if ((opts & OPENSSL_INIT_NO_LOAD_CRYPTO_STRINGS)
-            && !RUN_ONCE_ALT(&load_crypto_strings,
-                             ossl_init_no_load_crypto_strings,
-                             ossl_init_load_crypto_strings))
-        return 0;
-
-    if ((opts & OPENSSL_INIT_LOAD_CRYPTO_STRINGS)
-            && !RUN_ONCE(&load_crypto_strings, ossl_init_load_crypto_strings))
+    if ((opts & OPENSSL_INIT_NO_LOAD_CRYPTO_STRINGS)) {
+        if (!RUN_ONCE_ALT(&load_crypto_strings,
+                          ossl_init_no_load_crypto_strings,
+                          ossl_init_load_crypto_strings))
+            return 0;
+    } else if (!RUN_ONCE(&load_crypto_strings, ossl_init_load_crypto_strings)) {
         return 0;
+    }
 
     if ((opts & OPENSSL_INIT_NO_ADD_ALL_CIPHERS)
             && !RUN_ONCE_ALT(&add_all_ciphers, ossl_init_no_add_all_ciphers,
```

Reply:
But not again that we could also place `OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL)` in the functions that are used to retrieve strong values when exploring the error stack, i.e. as late as possible.

Reply:
I.e. this:

``` diff
diff --git a/crypto/err/err.c b/crypto/err/err.c
index c558bfa62b..d7a0c1ede8 100644
--- a/crypto/err/err.c
+++ b/crypto/err/err.c
@@ -589,6 +589,8 @@ void ERR_error_string_n(unsigned long e, char *buf, size_t len)
     if (len == 0)
         return;
 
+    OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL);
+
     l = ERR_GET_LIB(e);
     ls = ERR_lib_error_string(e);
     if (ls == NULL) {
```

Reply:
The more I think about it, the more I prefer the `crypto/init.c` diff...

Reply:
In the end, I chose the latter.  That seems like the sane way to avoid lockups.